Skip enumerating keys when cluster name is empty

tcsc opened the pull request:
Addresses Issue #5774

Prior to this change key enumeration could fail with an error if the `cluster` value in the `tsh` config was missing, which is possible when a post-v6.0 `tsh` reads a `~/.tsh` directory created by a pre-v6.0 `tsh`. This would ultimately cause the key enumeration code to search the wrong directory for keys, resulting in an attempt to read a directory as a key file, and failing.

This patch adds detection for an empty cluster name, and aborts the key enumeration without error if found.

tcsc committed:
Skip directories when enumerating keys

Addresses Issue #5774

If there is a spurious directory in a local keystore directory, `tsh`
will fail when enumerating keys. This is caused by the key enumeration
code attempting to read the directory lioke a key file and failing.

This patch makes the `*-db` and `*-kube` key enumeration skip any
directories it encounters in order for the enumeration to complete.

tcsc committed:
Typo fix

russjones commented:
@Joerger @fspmarshall @r0mant Can you review?

Joerger commented on lib/client/keystore_test.go:
```diff
@@ -80,6 +81,29 @@ func TestListKeys(t *testing.T) {
+	spuriousDirPath := path.Join(p, user+"-db", "root", "aaa-should-not-be-here")
+	require.NoError(t, os.MkdirAll(spuriousDirPath, 0700))
+
+	// When I attempt to enumerate the DB keys
```
nit:
```suggestion
	// When attempting to enumerate the DB keys
```

tcsc replied:
Why? (Not trying to be argumentative, just want to understand the rationale as I've always seen Given/When/Expect test info written in the first person)

russjones commented:
@tcsc Do you mean by partially addresses by the way? Is another PR coming?

tcsc commented:
> @tcsc Do you mean by partially addresses by the way? Is another PR coming?

Hmm. Poor choice of words on my part. Will update commit message.

tcsc commented:
Moved target branch from v6 to master as per [RFD-0012](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md).

tcsc committed:
Post-rebase test fixup

r0mant reviewed (commented):
Do we know how this might have happened? I.e. how could it happen that there's a directory where there should be a file?

The fix sort of looks fine to me but my main worry is that it's only fixing a very narrow scenario which doesn't seem to be possible under normal circumstances. And Ev said he never touched his ~/.tsh directory manually.

r0mant commented:
@awly You were refactoring K8s stuff a while ago - do you recall by chance if we ever used to store k8s cert as a file with the same name which we expect the `*-kube` directory to have now? This would explain how this might've happened - i.e. Ev was logged into a K8s cluster with an older version of tsh, then upgraded and our logout stumbled upon that old file.

That would also mean that db certs don't need this fix. Prob. could also tell by inspecting Ev's attached .tsh dir contents.

awly commented:
@r0mant see the 5.0 k8s cert refactoring changes: https://github.com/gravitational/teleport/pull/4769/files#diff-73e9b8ab27fb196aaaba6ba217f2179440b1febf016515621751a78a877a0e66

Previously, we used `foo-x509.pem` - the only TLS cert in `~/.tsh` which was used for everything (auth server and k8s). In 5.0, I added `foo-kube/` directory with per-cluster certs. There was no old `foo-kube` file.

Regardless of our changes, we shouldn't assume a pristine `~/.tsh` directory without weird files/modifications. We can't guarantee that the user or another program won't mess something up in there. The code should deal with any unexpected contents.

r0mant commented:
@tcsc @russjones So I did a bit of debugging and I'm pretty sure this fix is not what we want. I copied Ev's .tsh directory to my computer and this is what it looks like:

```
➜  e git:(3ec91b6) tree ~/.tsh/
/Users/r0mant/.tsh/
├── asteroid-sun.teleport.sh.yaml
├── current-profile
├── keys
│   └── asteroid-sun.teleport.sh
│       ├── certs.pem
│       ├── kontsevoy
│       ├── kontsevoy-cert.pub
│       ├── kontsevoy-kube
│       │   └── asteroid-sun.teleport.sh
│       ├── kontsevoy-x509.pem
│       └── kontsevoy.pub
├── known_hosts
└── profile -> asteroid-sun.teleport.sh.yaml

4 directories, 9 files
```

This is **correct** structure. There are no directories that should not be there - i.e. `~/.tsh/keys/asteroid-sun.teleport.sh/kontsevoy-kube/asteroid-sun.teleport.sh` is expected to be there and it's ok that it's empty.

Now, when you try to run any tsh command, it indeed fails:

```
➜  e git:(3ec91b6) build/tsh status --debug
ERRO [KEYSTORE]  read /Users/r0mant/.tsh/keys/asteroid-sun.teleport.sh/kontsevoy-kube/asteroid-sun.teleport.sh: is a directory client/keystore.go:333

ERROR REPORT:
Original Error: *os.PathError read /Users/r0mant/.tsh/keys/asteroid-sun.teleport.sh/kontsevoy-kube/asteroid-sun.teleport.sh: is a directory
Stack Trace:
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/lib/client/keystore.go:404 github.com/gravitational/teleport/lib/client.withKubeCerts.getKey
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/lib/client/keystore.go:332 github.com/gravitational/teleport/lib/client.(*FSLocalKeyStore).GetKey
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/lib/client/api.go:501 github.com/gravitational/teleport/lib/client.readProfile
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/lib/client/api.go:711 github.com/gravitational/teleport/lib/client.Status
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:1835 main.onStatus
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:515 main.Run
	/Users/r0mant/code/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:238 main.main
	/usr/local/Cellar/go@1.15/1.15.8/libexec/src/runtime/proc.go:204 runtime.main
	/usr/local/Cellar/go@1.15/1.15.8/libexec/src/runtime/asm_amd64.s:1374 runtime.goexit
User Message: read /Users/r0mant/.tsh/keys/asteroid-sun.teleport.sh/kontsevoy-kube/asteroid-sun.teleport.sh: is a directory
```

It fails here:

https://github.com/gravitational/teleport/blob/v6.0.2/lib/client/keystore.go#L390

and the reason is because the ReadDir call above (https://github.com/gravitational/teleport/blob/v6.0.2/lib/client/keystore.go#L382) actually enumerates this directory:

`~/.tsh/keys/asteroid-sun.teleport.sh/kontsevoy-kube`

instead of this one:

`~/.tsh/keys/asteroid-sun.teleport.sh/kontsevoy-kube/asteroid-sun.teleport.sh`

According to the code, this can happen when `o.teleportClusterName` is empty:

https://github.com/gravitational/teleport/blob/v6.0.2/lib/client/keystore.go#L381

And indeed, looking in Ev's profile file it's not there:

```
➜  e git:(3ec91b6) cat ~/.tsh/profile
web_proxy_addr: asteroid-sun.teleport.sh:443
ssh_proxy_addr: asteroid-sun.teleport.sh:3023
kube_proxy_addr: asteroid-sun.teleport.sh:3026
user: kontsevoy
```

Comparing with my own profile, notice the `cluster` field:

```
➜  e git:(3ec91b6) cat ~/.tsh/root.gravitational.io.yaml
web_proxy_addr: root.gravitational.io:3080
ssh_proxy_addr: root.gravitational.io:3023
mysql_proxy_addr: root.gravitational.io:3036
user: r0mant
cluster: root
```

I don't know whether this changed at some point, but the actual issue here is that in Ev's directory the "cluster" is empty, not that there are some extra directories.

If we want to handle this scenario, I think we should rather handle the case where `o.teleportClusterName` is empty gracefully.

r0mant reviewed (changes requested):
See my comment above ^

tcsc committed:
Address review comments

It turns out that the underlying issue is not that there are spurious sub-
directories in the `~/.tsh` dir; it's that the config is missing a `cluster`
value, which ends up causing the key enumeration code to enumerate the
parent directory rather than the expected subdirectory.

This change
 - removes the directory-detection code previously added
 - adds detection of an empty `teleportClusterName` value and aborts the
   key search if found.
 - updates tests to match new behaviour

tcsc commented:
> See my comment above ^

Done

r0mant commented on lib/client/keystore_test.go:
```diff
@@ -80,6 +80,24 @@ func TestListKeys(t *testing.T) {
+
+	// When I attempt to enumerate the DB keys with an empty teleport
+	// cluster name
+	_, err := s.store.GetKey(host, user, WithDBCerts("", ""))
```
I'd add `WithKubeCerts("", "")` as well to make sure both paths are covered by the test.

tcsc replied:
Done

russjones reviewed (approved):
Bot.

tcsc committed:
Adding kube cert enumeration to tests
